apps/labs/posts/scikit-lego-narwhals.md

```diff
@@ -0,0 +1,101 @@
+---
+title: 'How Narwhals and scikit-lego came together to enable dataframe-agnosticism'
+authors: [marco-gorelli]
+published: May 26, 2024
+description: 'And how your library can become dataframe-agnostic too'
+category: [PyData ecosystem]
+---
+
+# How scikit-lego became dataframe-agnostic
+
+Scikit-lego is a Python project which provides extra building blocks on top of scikit-learn.
+It provides some extra funcionality such as `TimeGapSplit` and `add_lags` which may be considered
+too niche for scikit-learn itself. These functions have always only ever supported pandas...up until now!
+
+As of scikit-lego 0.9.0, you can also pass dataframes from Polars, Modin, cuDF, and in principle a whole host
+of other dataframes.
+But, how does it work? Why did they use Narwhals?
+Why don't they just
+convert to pandas internally on the user's behalf?
+
+We'll start by answering these questions, and will end with some hopes for the future.
+
+## How does it work?
+
+Let's take a look at `sklego.pandas_utils.add_lags`. The code before version 0.9.0 did something like this:
+
+- check if the input is NumPy or a pandas dataframe
+- if it's NumPy, then perform array operations
+- if it's pandas, then perform pandas dataframe operations
+
+In order to support Polars too, one possible solution could have been do add a third step to
+the above:
+
+- if it's Polars, then perform Polars dataframe operations
+
+This looks simple, but it actually opens up a rabbit-hole of extra steps, such as:
```

Comment:
Quibble: standard usage is "rabbit hole", even though you're right that we usually hyphenate when using a noun as an adjective!